### PR TITLE
[core] Implements Session pool for the runtime env agent client.

### LINKED ISSUE #38662: Release test stress_test_many_runtime_envs.aws failed
Release test **stress_test_many_runtime_envs.aws** failed. See https://buildkite.com/ray-project/release-tests-branch/builds/2057#018a1723-e524-4ddf-aab3-3976c3ae116d for more details.

Managed by OSS Test Policy

### PR DESCRIPTION
**Pull request by rynewang**
Previously we have a *infinite* thread pool in the runtime env agent client: every time a call is needed, we create a new session with new TCP port to talk to the agent. This means on high load we can run out of the ports. This PR implements a simple session pool to limit the concurrency and queue up the overflowing requests.

Also, we had a test_many_runtime_envs release test that creates lots of working_dirs and run tasks. However it did not work because each time we updated files in the dir and use it as working_dir, but the path stays the same so Ray thinks it's the same runtime env. Changed it to test on env vars.

Fixes #38662.

### CODE REVIEW

**Comment by rynewang** [comment]
Release tests passing: https://buildkite.com/ray-project/release-tests-pr/builds/56672#018b5e04-c3db-4f2c-b562-c80c926a3fc0

After this PR I will make another PR to mark them stable & daily again.

File: src/ray/raylet/runtime_env_agent_client.cc [hunk 1]
```diff
@@ -182,6 +189,59 @@ class Session : public std::enable_shared_from_this<Session> {
...
+};
+
+// A pool of sessions with a fixed max concurrency. Each session can handle 1 concurrent
+// request.
+// Users can post a session to this pool, but should not *run* them. The Session pool runs
+// them if the concurrency is not exceeding the desired value.
+//
+// Once a session is done, remove it from the pool.
+//
+// NOT thread safe: if the methods or the fallbacks are invoked in different threads, the
```
Comment (rkooo567) [inline src/ray/raylet/runtime_env_agent_client.cc:212]: Should we add some sort of check in the constructor that checks the main thread id?

File: src/ray/raylet/runtime_env_agent_client_test.cc [hunk 2]
```diff
@@ -492,6 +504,79 @@ TEST(RuntimeEnvAgentClientTest, DeleteRuntimeEnvIfPossibleRetriesOnServerNotStar
...
+                                            /*agent_manager_retry_interval_ms=*/100);
+
+  size_t called_times = 0;
+  auto callback = [&](bool successful) {
+    ASSERT_TRUE(successful);
+    called_times += 1;
+  };
+
+  for (int i = 0; i < 100; ++i) {
+    client->DeleteRuntimeEnvIfPossible("serialized_runtime_env", callback);
```
Comment (rkooo567) [inline src/ray/raylet/runtime_env_agent_client_test.cc:571]: Can you add a test case that has some failures too? not just the successful case

**Review by rkooo567 — APPROVED** [review]
This is so much better! One minor comment for testing

File: src/ray/raylet/runtime_env_agent_client.cc (hunk 1 above)
Comment (rynewang) [reply src/ray/raylet/runtime_env_agent_client.cc:212]: If the user sticks to a single boost io_context this should be fine. If not we will have to use boost strand to limit it. But today I don't think we have more than 1 io_context anywhere.

File: src/ray/raylet/runtime_env_agent_client_test.cc (hunk 2 above)
Comment (rynewang) [reply src/ray/raylet/runtime_env_agent_client_test.cc:587]: done.

**Comment by rkooo567** [comment]
Screen Shot 2023-10-25 at 2 31 45 PM

need to fix this before merging it

**Comment by rkooo567** [comment]
@rynewang please lmk if you have different opinion!

**Comment by rynewang** [comment]
@rkooo567 you mean we don't cherry pick to *2.8* right? sgtm

**Pull request merged**